core/iwasm/fast-jit/fe/jit_emit_memory.c

```diff
@@ -765,27 +765,131 @@ bool
 jit_compile_op_atomic_rmw(JitCompContext *cc, uint8 atomic_op, uint8 op_type,
                           uint32 align, uint32 offset, uint32 bytes)
 {
+    assert(false && "atomic.rmw not implemented");
     return false;
 }
 
 bool
 jit_compile_op_atomic_cmpxchg(JitCompContext *cc, uint8 op_type, uint32 align,
                               uint32 offset, uint32 bytes)
 {
+    assert(op_type == VALUE_TYPE_I32
+           && "i64.atomic.rmw.cmpxchg not implemented");
+    assert(bytes == 4 && "only bytes=4 is implemented");
+
+    JitReg sval, expect, addr, memory_data, readv, offset1, new_val, res;
+    JitReg args[1] = { 0 };
+
+    POP_I32(sval);
+    POP_I32(expect);
+    POP_I32(addr);
+
+    offset1 = check_and_seek(cc, addr, offset, bytes);
+    if (!offset1) {
+        goto fail;
+    }
+    memory_data = get_memory_data_reg(cc->jit_frame, 0);
+
+    readv = jit_cc_new_reg_I32(cc);
+    new_val = jit_cc_new_reg_I32(cc);
+
+    // Memory mutex lock
+    res = jit_cc_new_reg_I32(cc);
+    args[0] = get_module_inst_reg(cc->jit_frame);
+    if (!jit_emit_callnative(cc, wasm_runtime_mem_lock, res, args,
+                             sizeof(args) / sizeof(args[0])))
+        goto fail;
+
+    GEN_INSN(LDU32, readv, memory_data, offset1);
+
+    // new_val = (readv == expect) ? sval : readv
+    GEN_INSN(CMP, cc->cmp_reg, readv, expect);
+    GEN_INSN(SELECTEQ, new_val, cc->cmp_reg, sval, readv);
+    GEN_INSN(STI32, new_val, memory_data, offset1);
+
+    // Memory mutex unlock
+    if (!jit_emit_callnative(cc, wasm_runtime_mem_unlock, res, args,
+                             sizeof(args) / sizeof(args[0])))
+        goto fail;
+
+    PUSH_I32(readv);
+    return true;
+fail:
     return false;
 }
 
 bool
 jit_compile_op_atomic_wait(JitCompContext *cc, uint8 op_type, uint32 align,
```

Comment:
Should we also check the alignment of the address in all those functions?

> Unlike normal memory accesses, misaligned atomic accesses trap. For non-atomic accesses on shared linear memory, misaligned accesses do not trap.

Reply:
yes, I need to add that part too, similarly to what is done in the classic interpreter https://github.com/bytecodealliance/wasm-micro-runtime/blob/cadf9d0ad36ec12e2a1cab4edf5f0dfb9bf84de0/core/iwasm/interpreter/wasm_interp_classic.c#L67